Framework app control panel

ptomato opened the pull request:
This is the whole of the control panel except for some sound packs which I haven't curated yet. (We still need two click sound packs and three hover sound packs, of 6 sounds each.)

The code view is not implemented yet.

https://phabricator.endlessm.com/T23625

ptomato committed:
framework: Add framework app hacking toolbox

This adds a control panel for hacking eos-knowledge-lib-based apps.
Currently they need to be whitelisted, and we start with the only
whitelisted app being com.endlessm.dinosaurs.en.

The UI elements in the control panel should work at this point, but
don't do anything yet, even when flipping back to the app.

https://phabricator.endlessm.com/T23626

ptomato committed:
Apply framework controls to framework apps

Adds the code to take the current state of the model, and generate
app.json, overrides.css files, and GResource files; and relaunch the app
using the Restart API from eos-knowledge-lib.

We keep all the custom modules in a separate GResource file and pass it
directly when we relaunch the app.

https://phabricator.endlessm.com/T23626

ptomato committed:
Copy some modules from modular framework

Due to a GJS bug (https://gitlab.gnome.org/GNOME/gjs/issues/97) where it's
impossible to inherit from a template class in GJS, we need to reimplement
some framework modules where otherwise we would have just inherited from
them.

This copies the exact files for Banner.Dynamic, Banner.Search,
Card.Default, and Card.Title from the eos-knowledge-lib repo.

This commit can mostly be skipped in review, since the code is all
copied.

https://phabricator.endlessm.com/T23626

ptomato committed:
Lint copied modules from modular framework

Make minimal changes to the modules copied from the modular framework to
get them to pass our linter checks in this repository. Additionally
remove the documentation comments since we are not building
documentation in this repo.

This commit can be reviewed briefly since it is mostly mechanical
changes.

https://phabricator.endlessm.com/T23626

ptomato committed:
Card.Default: Rewrite confusing code that bugs me

I could not figure out at all what this was doing. Rewrite it, and
hopefully I will remember to upstream it into eos-knowledge-lib.

https://phabricator.endlessm.com/T23626

cosimoc commented on src/framework/appWindow.js:
```diff
@@ -0,0 +1,24 @@
+var RaAppWindow = GObject.registerClass(class RaAppWindow extends ToolboxWindowBase {
+    _init(props = {}) {
+        super._init(props);
+        this.maximize();
```
I wouldn't do this here... The flip to hack code in the shell should already take care of synchronizing the toolbox window geometry with the original one.

cosimoc commented on src/framework/model.js:
```diff
@@ -0,0 +1,211 @@
+
+        if (value !== 'none') {
+            this._clickSound = 'none';
+            this.notify('click-sound');
```
Why can't a click sound and a hover sound be both set at the same time?

ptomato replied:
No fundamental reason, but this made it simpler to implement and I also thought it would make it difficult to play a melody on the piano, for example, if you were also getting hover sounds in between. I can remove that restriction, but if so we should create another issue for it.

cosimoc replied:
Sounds good

cosimoc commented on src/meson.build:
```diff
@@ -55,3 +55,5 @@ gnome.generate_gir(main_library, extra_args: ['--warn-all', '--warn-error'],
 
 install_headers(installed_headers, subdir: api_name)
+
+subdir('framework')
```
This should probably go in the previous commit?

ptomato replied:
There was no meson.build file needed in the framework directory until this commit — it's only needed to build the GResource file with the extra framework modules.

cosimoc commented on src/framework/textFilters.js:
```diff
@@ -0,0 +1,255 @@
+
+function zalgo(text) {
+    // Thanks to Michael S. Kaplan:
+    // http://blogs.msdn.com/b/michkap/archive/2006/02/17/533929.aspx
```
Page does not seem to exist :(

ptomato replied:
I see ... I looked it up on archive.org and it's really only tangentially related, so I'll just delete this.

cosimoc commented on src/framework/textFilters.js:
```diff
@@ -0,0 +1,255 @@
+
+        // Middle
+        // Put just one of the middle characters there, or it gets crowded
+        const middleLen = ZALGO_DIACRITICS.middle.length;
```
Should this not be length -1?

ptomato replied:
No, I don't think so - `Math.floor(Math.random() * middleLen)` will return an index from 0 to `middleLen - 1`, so it works out okay.

cosimoc commented on src/framework/model.js:
```diff
@@ -208,4 +227,86 @@ var RaModel = GObject.registerClass({
+        const gresource = tmpDir.get_child('hacking.gresource');
+
+        const logoInputStream = await logoResource.read_async(_PRIO, null);
+        const logoOutputStream = await logo.replace_async(null, false,
```
Couldn't you simply use `Gio.File.copy()` for this?

cosimoc commented on src/framework/model.js:
```diff
@@ -208,4 +227,86 @@ var RaModel = GObject.registerClass({
+        proc.init(null);
+        await proc.wait_check_async(null);
+
+        // Ah, geez, this is terrible
```
What is this doing exactly?

ptomato replied:
So, this is the only way I could find to open a file and get an FD using GJS ... it uses Gio.Subprocess to run the command `cat myfile`, and grabs the FD from its stdout pipe.

cosimoc commented on src/framework/banner/hackableSearch.js:
```diff
@@ -0,0 +1,52 @@
+// Copyright 2015 Endless Mobile, Inc.
```
2018

cosimoc reviewed (changes requested):
Looks pretty good! This is a first pass review, as there's quite a bit of code. @tchx84 or @xjuan can perhaps help with the next round...

cosimoc committed:
appWindow: enable flip-back action for framework windows

cosimoc committed:
app: don't close toolbox when flipping back

We'll still need this to stick around for further changes.

cosimoc committed:
app: delete window after we're done with it

ptomato commented:
OK, finally made the changes based on the above comments. I'll rebase the `controlpanel-cosimoc` branch on top of this.

ptomato committed:
Internal HackToolbox library

There was already infrastructure there for a library containing a DBus
skeleton from generated code, so this adds a non-generated header and
source file with a function that will open a file and give back an FD.

https://phabricator.endlessm.com/T23626

ptomato commented:
Added the internal C library for opening files.
